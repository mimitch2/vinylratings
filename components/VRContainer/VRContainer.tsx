import React, { useRef, useEffect } from 'react';
import {
    Animated,
    StyleSheet,
    View,
    RefreshControl,
    ViewStyle
} from 'react-native';
// import { useTheme } from '@react-navigation/native';
import { Layout } from '@ui-kitten/components';

import { VRLoading } from 'components';
import { FONTS, Theme, ThemeColors } from 'constants/index';

interface Props {
    children: React.ReactNode;
    scrollable?: boolean;
    isLoading?: boolean | undefined;
    animationDuration?: number;
    animationDelay?: number;
    startAnimation?: boolean;
    styleOverride?: ViewStyle;
    refreshing?: boolean;
    onRefresh?: any;
}

const VRContainer = ({
    children,
    scrollable = true,
    isLoading = false,
    animationDuration = 100,
    animationDelay = 50,
    startAnimation = false,
    styleOverride = {},
    refreshing = false,
    onRefresh = null
}: Props) => {
    const fadeIn = useRef(new Animated.Value(0)).current;

    const commonStyle = [
        {
            padding: 20,
            opacity: fadeIn
        },
        styleOverride
    ];

    useEffect(() => {
        if (startAnimation) {
            Animated.timing(fadeIn, {
                toValue: 1,
                useNativeDriver: true,
                duration: animationDuration,
                delay: animationDelay
            }).start();
        }
    }, [fadeIn, startAnimation, animationDuration, animationDelay]);

    if (isLoading) {
        return <VRLoading />;
    }

    return scrollable ? (
        <Layout
            style={[styles.view, { opacity: refreshing ? 0.6 : 1 }]}
            testID="scroll"
        >
            <Animated.ScrollView
                contentInsetAdjustmentBehavior="automatic"
                style={commonStyle}
                refreshControl={
                    onRefresh ? (
                        <RefreshControl
                            refreshing={refreshing}
                            onRefresh={onRefresh}
                            // colors={[colors.lightGrey]}
                            // tintColor={colors.lightGrey}
                        />
                    ) : undefined
                }
            >
                <View>{children}</View>
            </Animated.ScrollView>
        </Layout>
    ) : (
        <Layout style={styles.view} testID="view">
            <Animated.View style={commonStyle}>
                <View>{children}</View>
            </Animated.View>
        </Layout>
    );
};

const styles = StyleSheet.create({
    view: {
        flex: 1,
        fontFamily: FONTS.primary
    }
});

export default React.memo(VRContainer);
