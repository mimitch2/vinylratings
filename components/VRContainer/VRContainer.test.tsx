import React from 'react';
import { Text } from 'react-native';
import { waitFor } from '@testing-library/react-native';

import VRContainer from './VRContainer';
import { renderWithProvider } from 'test';

describe('VRContainer', () => {
    it('should render loading', async () => {
        const { getByTestId } = renderWithProvider(
            <VRContainer isLoading>
                <Text>Testing this yo</Text>
            </VRContainer>
        );

        await waitFor(() => {
            expect(getByTestId('loading')).toBeTruthy();
        });
    });
    it('should render correctly for scrollable', () => {
        const { getByTestId } = renderWithProvider(
            <VRContainer startAnimation>
                <Text>Testing this yo</Text>
            </VRContainer>
        );

        expect(getByTestId('scroll')).toBeTruthy();
    });

    it('should render correctly for non-scrollable', () => {
        const { getByTestId } = renderWithProvider(
            <VRContainer startAnimation scrollable={false}>
                <Text>Testing this yo</Text>
            </VRContainer>
        );

        expect(getByTestId('view')).toBeTruthy();
    });

    it('should render refresh', () => {
        const view = renderWithProvider(
            <VRContainer onRefresh={() => {}} refreshing>
                <Text>Testing this yo</Text>
            </VRContainer>
        );

        expect(view).toMatchSnapshot();
    });
});
