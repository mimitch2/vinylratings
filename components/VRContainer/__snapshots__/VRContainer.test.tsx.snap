// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VRContainer should render refresh 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "rgb(242, 242, 242)",
        "flex": 1,
        "fontFamily": "ArchivoNarrow-Regular",
      },
      {
        "opacity": 0.6,
      },
    ]
  }
  testID="scroll"
>
  <RCTScrollView
    collapsable={false}
    contentInsetAdjustmentBehavior="automatic"
    refreshControl={
      <RefreshControlMock
        colors={
          [
            undefined,
          ]
        }
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    scrollEventThrottle={0.0001}
    style={
      {
        "opacity": 0,
        "padding": 20,
      }
    }
  >
    <RCTRefreshControl />
    <View>
      <View>
        <Text>
          Testing this yo
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
